### ISSUE TITLE
OIDC: Handle `audience` query parameter

### ISSUE THREAD

**Issue by nicholascioli** [sebadob/rauthy#650]
I noticed when making requests with [oauth2c](https://github.com/cloudentity/oauth2c) and specifying custom audiences that the returned token's `aud` field is always set to the client ID.

After looking at the query parameters supported by the [OIDC login flow](https://github.com/sebadob/rauthy/blob/85ad0959380860e1d912b52808a6a2ad20df18f4/src/api_types/src/oidc.rs#L34) I noticed that it doesn't read the audience query parameter at all.

Is this something that you would consider incorporating into rauthy? If so, I can take a stab at making a PR, but since I'm not super familiar with the security implications of custom audiences, I figured that I'd ask.

Thanks again!

**Comment by sebadob** [comment]
### tl;dr

You must not accept a custom `aud` from any client at all, it would break your whole security.

### Long Version

You can see all valid OIDC request parameters [here in the RFC](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest). The `aud` does not exist, for very good reason.

When clients validate a given JWT Token, they of course validate things like the expiry, the issuer it came from, the signature, and so on. One of the most important factors for validation is often forgotten: Make sure the token is actually for you!

Let's say you have 2 clients called `batman` and `joker`. Both clients are valid, exist on your IdP, do have their own setup, config, secrets, mappings, and everything. If you log in to `batman` and are redirected to the Batman Application, it will make sure that the token is valid in all terms, of course. One of the most important steps in validation though is the `aud` claim. It must check that the `aud` states it's actually a token for `batman` and not something else. This is (by default) done via the `aud` claim, which in 99% of cases simply matches the `client_id`. In fact, this is even a MUST from the [spec](https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation).

If the sneaky `joker` comes in, logs you in in a totally valid way, and let's assume he would be able to pass an `aud` param during the authorization containing `batman` as a custom audience. If `joker` now even has different roles and groups mappings, for instance a user is an `admin` on `joker` but not on `batman`, the role mappings would look different in the token as well, contain different information, and so on. The crucial part is, if `joker` was able to request a 100% valid token containing a `batman` audience, he could use that token on the `batman` API and would have validated access, because `batman` thinks the token is for him, since `batman`, which is his `client_id`, is in the  `aud` claim.

The whole story gets even worse when dynamic or ephemeral clients are enabled. Then they don't even need to be pre-registered on the IdP to request a valida token, but with an `aud` claim matching another existing `client_id` to access its API.

Theoretically, you could do something like custom audiences, and make it somewhat secure. For instance, make sure that the `client_id` is always the very first claim in the array. But, this is super dangerous, because you cannot know, how client implementations validate the token. In fact, most will only make a comparison and check, if the `client_id` exists in the `aud` claim (some don't even check `aud` which is another story).

Currently, there is only one situation, where you would have more than only the `client_id` in the `aud` claim, and this is when you have enabled Solid OIDC. In this case, Rauthy will add the mandatory `solid` claim to `aud`.

Long story short, it would theoretically be possible to allow something like this, but this is very dangerous territory and I did not see any reason so far, why you would need something. Which brings me to my question: 
What do you want to achieve with a custom `aud`?

**Comment by nicholascioli** [comment]
Hmm, yeah all of that makes sense. The main thing that I was trying to figure out is how to have a server validate that an ID token of a client is valid. Like you said, most implementations verify that the audience of the ID token matches the client ID, but how does that work when you have multiple clients, each with their own client ID? Is the idea to not operate on ID tokens on the server side at all? I was hoping to piggy-back off of the ID token when doing authn / authz in order to avoid having to have a completely duplicate set of users, but if that is not the typical use case then I'm not sure what the best way to move forward is.

Solid OIDC (from the small amount that I've seen) seems like maybe it addresses this multi-client, single server token validation, but maybe all that is needed is to have the server call the token introspection endpoint to validate a token and then extract claims / roles / etc.?

Sorry if these questions seem very basic; I'm very new to the OIDC world.

**Comment by sebadob** [comment]
The `id_token` should never be used for any authn / authz. The `aud` also does not only exist in the `id_token`.
With most client's check the `aud`, I meant they are doing only that. Others do more stuff (like the non-default `azp` for instance) and the ones that don't validate `aud` at all are actually broken, since this is mandatory. JWT validation has a lot of footguns and tbh, it's a bad design from that standpoint. It's probably only used because it's wide spread, but that does not mean it's good. 
It's very rare that an application accepts multiple `aud`iences. Usually you have 1 client for 1 app / backend, that's it.

It's probably a good idea to read the whole [OIDC 1.0 Core RFC](https://openid.net/specs/openid-connect-core-1_0.html) once to understand all moving bits an pieces.
